### PR TITLE
OHRM5X-559: Attendance report end point added

### PR DESCRIPTION
**Pull request by VageeshOrangeHRM**
This PR Contains following changes and fixes:
1. Attendance Report End point added for Table Structure and Data.
2. Data group and Role permission added.
3. Unit tests and integration test added.
4. OHRM5X-1072 - [Time - Employee timesheets] Timesheet comment fields should not be a required field] issue fixed.

### CODE REVIEW

File: symfony/lib/orm/TimeDiff.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+     * @var string
+     */
+    public $dateTime2;
+
+    /**
+     * @var string
+     */
+    public $type;
+
+    public function parse(Parser $parser)
```
Comment (RajithaKumara) [inline symfony/lib/orm/TimeDiff.php:44]: doc comment

File: symfony/lib/orm/TimeDiff.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        $parser->match(Lexer::T_OPEN_PARENTHESIS);
+        $this->dateTime1 = $parser->ArithmeticPrimary();
+        $parser->match(Lexer::T_COMMA);
+        $this->dateTime2 = $parser->ArithmeticPrimary();
+        $parser->match(Lexer::T_COMMA);
+        $this->type = $parser->Literal()->value;
+        $parser->match(Lexer::T_CLOSE_PARENTHESIS);
+    }
+
+    public function getSql(SqlWalker $sqlWalker)
```
Comment (RajithaKumara) [inline symfony/lib/orm/TimeDiff.php:56]: doc comment, return type

File: symfony/plugins/orangehrmAttendancePlugin/Dao/AttendanceDao.php [hunk 3]
```diff
@@ -626,4 +628,89 @@ public function getLatestAttendanceRecordByEmployeeNumber(int $employeeNumber):
 
         return $q->getQuery()->getOneOrNullResult();
     }
+
+    /**
+     * @param AttendanceReportSearchFilterParams $attendanceReportSearchFilterParams
+     * @return array
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAttendancePlugin/Dao/AttendanceDao.php:634]: you can add key list of associate array, or example

File: symfony/plugins/orangehrmTimePlugin/Api/TimeReportAPI.php [hunk 4]
```diff
@@ -33,6 +34,7 @@ class TimeReportAPI extends ReportAPI
         //activity_detailed -> detailed report of project report (Break down overview of project report)
         'activity_detailed' => ProjectActivityReport::class,
         'employee' => EmployeeReport::class,
+        'attendance' => AttendanceReport::class,
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Api/TimeReportAPI.php:37]: is this part of time plugin or attendance plugin

File: symfony/plugins/orangehrmTimePlugin/Report/AttendanceReport.php [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+    /**
+     * @inheritDoc
+     */
+    public function getValidationRule(EndpointProxy $endpoint): ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            $endpoint->getValidationDecorator()->notRequiredParamRule(
+                new ParamRule(
+                    CommonParams::PARAMETER_EMP_NUMBER,
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Report/AttendanceReport.php:124]: in this class you have `self::FILTER_EMP_NUMBER`

File: symfony/plugins/orangehrmTimePlugin/Report/AttendanceReport.php [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+            )
+        );
+    }
+
+    /**
+     * @return Header
+     */
+    public function getHeaderDefinition(): HeaderDefinition
+    {
+        // TODO need to finalize
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Report/AttendanceReport.php:177]: remove TODO

File: symfony/plugins/orangehrmTimePlugin/Report/AttendanceReportData.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        $employeeAttendanceRecords = $this->getAttendanceService()
+            ->getAttendanceDao()
+            ->getAttendanceReportCriteriaList($this->filterParams);
+
+        $result = [];
+        foreach ($employeeAttendanceRecords as $employeeAttendanceRecord) {
+            $termination = $employeeAttendanceRecord['termination'];
+            $result[] = [
+                AttendanceReport::PARAMETER_EMPLOYEE_NAME => $termination === null ? $employeeAttendanceRecord['fullName'] : $employeeAttendanceRecord['fullName'] . ' (Past employee)',
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Report/AttendanceReportData.php:57]: :+1:

File: symfony/plugins/orangehrmTimePlugin/test/Api/AttendanceReportAPITest.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+use OrangeHRM\Config\Config;
+use OrangeHRM\Framework\Services;
+use OrangeHRM\Tests\Util\EndpointIntegrationTestCase;
+use OrangeHRM\Tests\Util\Integration\TestCaseParams;
+use OrangeHRM\Tests\Util\TestDataService;
+use OrangeHRM\Time\Api\TimeReportAPI;
+
+/**
+ * @group Time
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/test/Api/AttendanceReportAPITest.php:30]: ??? :thinking: 
```suggestion
 * @group Attendance
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]
Good job :tada:

File: symfony/lib/orm/TimeDiff.php (hunk 1 above)
Comment (VageeshOrangeHRM) [reply symfony/lib/orm/TimeDiff.php]: done

File: symfony/lib/orm/TimeDiff.php (hunk 2 above)
Comment (VageeshOrangeHRM) [reply symfony/lib/orm/TimeDiff.php]: done

File: symfony/plugins/orangehrmAttendancePlugin/Dao/AttendanceDao.php (hunk 3 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmAttendancePlugin/Dao/AttendanceDao.php:634]: done

File: symfony/plugins/orangehrmTimePlugin/Api/TimeReportAPI.php (hunk 4 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Api/TimeReportAPI.php:37]: time?????

File: symfony/plugins/orangehrmTimePlugin/Report/AttendanceReport.php (hunk 5 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Report/AttendanceReport.php]: done

File: symfony/plugins/orangehrmTimePlugin/Report/AttendanceReport.php (hunk 6 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Report/AttendanceReport.php]: done

File: symfony/plugins/orangehrmTimePlugin/Report/AttendanceReportData.php (hunk 7 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Report/AttendanceReportData.php:57]: done

File: symfony/plugins/orangehrmTimePlugin/test/Api/AttendanceReportAPITest.php (hunk 8 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/test/Api/AttendanceReportAPITest.php:30]: time????

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**